### ISSUE TITLE
Migration from v1 to v2 not working, intercept not working (minimal example for GCP inside)

### ISSUE THREAD

**Issue by craftey** [telepresenceio/telepresence#2798]
I am trying to migrate a telepresence v1 call that swaps a deployment to telepresence v2. I have a java application in a docker container that I want to boot locally and have it connect to all the servers that my remote pod can connect to.

I could reduce the issue to a minimal example that you can reproduce. I read [here](https://github.com/telepresenceio/telepresence/issues/2758#issuecomment-1261581609) and [there](https://datawire-oss.slack.com/archives/CAUBBJSQZ/p1661965493444989?thread_ts=1661549637.194529&cid=CAUBBJSQZ) that it might be a firewall issue with GCP. Though, I was not able to fix the issue by setting a firewall rule.
I ran the minimal example without a VPN involved.
Tests were done with **telepresence version 2.7.6**.

It would be nice if anybody of the telepresence team could reproduce below issue:

**Minimal example in GCP:**

Create k8s-cluster and deployment in GCP according to this howto: 
https://cloud.google.com/kubernetes-engine/docs/deploy-app-cluster#standard
You can adjust the zone to your likings.
```bash
gcloud container clusters create k8s-test-cluster --num-nodes=1 --zone=europe-west3-a
gcloud container clusters get-credentials k8s-test-cluster --zone=europe-west3-a
kubectl create namespace test
kubectl create deployment hello-server --image=us-docker.pkg.dev/google-samples/containers/gke/hello-app:1.0 -n test
kubectl rollout status deployment/hello-server --namespace test
kubectl expose deployment hello-server --type LoadBalancer --port 80 --target-port 8080 -n test
```
Then run the following commands:

```bash
telepresence helm install

helm list -n ambassador 
#NAME           	NAMESPACE 	REVISION	UPDATED                              	STATUS  	CHART             	APP VERSION
#traffic-manager	ambassador	1       	2022-09-29 13:15:41.374646 +0200 CEST	deployed	telepresence-2.7.6	2.7.6      

# 169.254.169.254 is the ip of metadata.google.internal
telepresence intercept hello-server \
  --namespace test \
  --docker-run -- \
  --rm \
  curlimages/curl:7.85.0 \
  -v --connect-timeout 10 -H 'Metadata-Flavor:Google' \
  http://169.254.169.254/computeMetadata/v1/project/project-id

# same server but with DNS involved
# http://metadata.google.internal/computeMetadata/v1/project/project-id
```

I would expect the GCP project id as response from the curl command.
But instead **I get a timeout.**

```bash
# checking agent
telepresence list -n test
hello-server: ready to intercept (traffic-agent already installed)
```

I derived the command from [telepresence migration docu](https://www.getambassador.io/docs/telepresence/latest/install/migrate-from-legacy/).

If I **switch to telepresence v1** on my Mac it just works as expected:
```bash
brew unlink telepresence ; brew link telepresence-legacy # tele1

telepresence --swap-deployment hello-server \
  --namespace test \
  --docker-run \
  --rm \
  curlimages/curl:7.85.0 \
  -v --connect-timeout 10 -H 'Metadata-Flavor:Google' \
  http://169.254.169.254/computeMetadata/v1/project/project-id
```
I get my project id printed out.
This type of command is what I used in the past to run my local java app in docker container with telepresence v1.

Note, I also tried with legacy telepresence v1 the following commands:
```bash
telepresence --run curl -v --connect-timeout 10 -H 'Metadata-Flavor:Google' \
  http://169.254.169.254/computeMetadata/v1/project/project-id

telepresence --swap-deployment hello-server --namespace test --run \
  curl -v --connect-timeout 10 -H 'Metadata-Flavor:Google' \
  http://169.254.169.254/computeMetadata/v1/project/project-id
```
Both timed out with legacy telepresence v1.
The corresponding telepresence v2 commands also failed with timeout.

**So my questions:**
1. Can you reproduce the issue?
2. Can I change something in the usage of telepresence to work around the issue?
3. If a firewall rule can fix the issue, how exactly would I do that in my minimal example?

Many thanks in advance,
craftey

**Cleanup experiment in GCP**
```bash
kubectl delete service hello-server -n test
gcloud container clusters delete k8s-test-cluster --zone=europe-west3-a
```

**Comment by thallgren** [comment]
Hey @craftey. I've noticed some timing problems when executing using `--docker-run` which will be fixed in a coming release. Not sure if they are applicable here, but one thing you could try to check this is.
```console
$ telepresence intercept hello-server  --namespace test
```
This will start the actual intercept. Then, once that is up and running and a few seconds have passed, start your docker container.
```console
$ docker run \
  --dns-search tel2-search \
  --env-file /tmp/envs.env \
  --rm \
  curlimages/curl:7.85.0 \
  -v --connect-timeout 10 -H 'Metadata-Flavor:Google' \
  http://169.254.169.254/computeMetadata/v1/project/project-id
```
You can leave the intercept with:
```console
$ telepresence leave hello-server-test
```
If that doesn't help, please try
```console
$ telepresence loglevel debug
<reproduce the problem>
$ telepresence gather-logs
```
and check the generated logs. Feel free to convey them here so that we can help analyze the problem.

**Comment by craftey** [comment]
Thanks @thallgren. 
I tried ur suggestion in my minimal setup. It was not successful.

```bash
$ telepresence intercept hello-server  --namespace test ; sleep 10
Using Deployment hello-server
intercepted
    Intercept name         : hello-server-test
    State                  : ACTIVE
    Workload kind          : Deployment
    Destination            : 127.0.0.1:8080
    Service Port Identifier: 80
    Volume Mount Point     : /var/folders/wc/34vwg6_n2yl6jmkv_pwqml7w0000gr/T/telfs-1967567784
    Intercepting           : all TCP requests
```
And then:
```
 docker run \
  --dns-search tel2-search \
  --rm \
  curlimages/curl:7.85.0 \
  -v --connect-timeout 10 \
  -H 'Metadata-Flavor:Google' \
  http://169.254.169.254/computeMetadata/v1/project/project-id
```
I think equivalent is just:
```
curl -v --connect-timeout 10 -H 'Metadata-Flavor:Google' \
  http://169.254.169.254/computeMetadata/v1/project/project-id
```
Both commands timed out.

I also tried
```
curl -v --connect-timeout 10 -ik https://kubernetes.default
```
At first this did not work, I got
```
curl: (6) Could not resolve host: kubernetes.default
```
I was a bit confused, because this works on my other cluster, where my original problem is, quite well.
Then I tested `kubernetes.default` with telepresence v1, it worked.
Then I learned from [here](https://github.com/telepresenceio/telepresence/issues/2746#issuecomment-1253126027) that I need to restart traffic-manager in such a case. 
@DanielYWoo  wrote:
> I guess the DNS initialization happens only when the traffic manager starts, if something is wrong then it's wrong forever.

So I did:
```
kubectl rollout restart -n ambassador deployment traffic-manager
```
This fixed connecting to `kubernetes.default`.
However, this would also work if I just ran `telepresence connect` without any active intercept.
I think, my traffic-manager got into this state, because to save money I shut down the nodepool of my example cluster when I do not use it. Upon activating the nodepool, I guess, traffic-manager can get into that state. I will retest this and create an extra ticket for this, if it is reproducible.

After I learned this, I retried everything above with `169.254.169.254` but still get the timeout.

**Comment by craftey** [comment]
Next, I will create logs, as you suggested @thallgren. But I am still interested if you can reproduce the issue.

**Comment by craftey** [comment]
```
$ telepresence quit
Telepresence Network disconnecting...done
Telepresence Traffic Manager disconnecting...done
$ telepresence loglevel debug
Connected to context gke_foo-cbdp-infraqa-core_europe-west3-a_k8s-test-cluster (https://xx.xx.xx.xx)
$ telepresence intercept hello-server  --namespace test ; sleep 10
Using Deployment hello-server
intercepted
    Intercept name         : hello-server-test
    State                  : ACTIVE
    Workload kind          : Deployment
    Destination            : 127.0.0.1:8080
    Service Port Identifier: 80
    Volume Mount Point     : /var/folders/wc/34vwg6_n2yl6jmkv_pwqml7w0000gr/T/telfs-4253709182
    Intercepting           : all TCP requests
$ curl -v --connect-timeout 10 -H 'Metadata-Flavor:Google' http://169.254.169.254/computeMetadata/v1/project/project-id
[...]
curl: (28) Connection timeout after 10002 ms
$ telepresence gather-logs
Logs have been exported to /Users/me/telepresence_logs.zip
```

[traffic-manager-866dc64459-82mqc.ambassador.log](https://github.com/telepresenceio/telepresence/files/9724801/traffic-manager-866dc64459-82mqc.ambassador.log)
[hello-server-6988df5dbf-f4ggp.test.log](https://github.com/telepresenceio/telepresence/files/9724802/hello-server-6988df5dbf-f4ggp.test.log)
[connector.log](https://github.com/telepresenceio/telepresence/files/9724818/connector.log)
[daemon.log](https://github.com/telepresenceio/telepresence/files/9724850/daemon.log)

Btw, in these logs I see many read error, connection refused and read canceled with UDP protocol. Is this an issue?
Also, when searching the logs I do not find `169.254.169.254` at all. Is that expected?

**Comment by thallgren** [comment]
By looking at the logs for the daemon and the traffic-manager, it seems that no DNS requests at all were made that were redirected to Telepresence, which isn't surprising, given that you actually use an IP address in the URL. So I think we can rule out that this is DNS related.

Telepresence adds two subnets to its VIF, which is responsible for routing traffic to the cluster. Those are visible in the daemon.log:
```
2022-10-06 09:51:42.5496 info    daemon/session/watch-cluster-info : Adding service subnet 10.64.0.0/20
2022-10-06 09:51:42.5497 info    daemon/session/watch-cluster-info : Adding pod subnet 10.60.0.0/24
```
You then try the IP `169.254.169.254` which, not surprisingly, fails. It's not an IP in your cluster.

**Comment by thallgren** [comment]
Perhaps this is a subnet that your cluster has access to, that you also want your local machine have access to? If so, please have a look at how to configure the VIF with additional subnets using [also-proxy](https://www.getambassador.io/docs/telepresence/latest/reference/config/#alsoproxy).

**Comment by craftey** [comment]
Thanks again @thallgren. I will try that and report back.
I understand that this is due to the agent approach of telepresence v2. In my real application I will have more than this subnet to add. Will see how that goes. 

One remark would be, that this is not as convenient as it was with swap-deployment from telepresence v1. Do you think a swap-like user experience can be brought back to v2?

**Comment by thallgren** [comment]
I do believe that you had to use `also-proxy` in v1 too, if you wanted access to non-cluster subnets. So perhaps there's more than one issue here?

**Comment by thallgren** [comment]
And no, this is not related to the agent. The `also-proxy` is needed even if you don't intercept. It's all about outbound networking and making IPs that aren't covered by the cluster's subnets available.

**Comment by craftey** [comment]
For my test cluster I added

```    
    extensions:
    - name: telepresence.io
      extension:
        also-proxy:
        - 169.254.169.254/32
``` 
to local kube config file within the cluster entry.

The `intercept --docker-run` command works now when using the ip.

But using the DNS name instantly fails:
```
$ telepresence intercept hello-server   --namespace test   --docker-run --   --rm   curlimages/curl:7.85.0   -v --connect-timeout 10   -H 'Metadata-Flavor:Google' http://metadata.google.internal/computeMetadata/v1/project/project-id
Connected to context gke_foo_cbdp-infraqa-core_europe-west3-a_k8s-test-cluster (https://xx.xx.xx.xx)
Using Deployment hello-server
intercepted
    Intercept name         : hello-server-test
    State                  : ACTIVE
    Workload kind          : Deployment
    Destination            : 127.0.0.1:8080
    Service Port Identifier: 80
    Volume Mount Point     : /var/folders/wc/34vwg6_n2yl6jmkv_pwqml7w0000gr/T/telfs-2932397200
    Intercepting           : all TCP requests
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* Could not resolve host: metadata.google.internal
* Closing connection 0
curl: (6) Could not resolve host: metadata.google.internal
telepresence: error: /usr/local/bin/docker docker run --dns-search tel2-search --env-file /var/folders/wc/34vwg6_n2yl6jmkv_pwqml7w0000gr/T/tel-3567480296.env --name intercept-hello-server-test-8080 -p 8080:8080 -v /var/folders/wc/34vwg6_n2yl6jmkv_pwqml7w0000gr/T/telfs-2932397200:/var/folders/wc/34vwg6_n2yl6jmkv_pwqml7w0000gr/T/telfs-2932397200 --rm curlimages/curl:7.85.0 -v --connect-timeout 10 -H Metadata-Flavor:Google http://metadata.google.internal/computeMetadata/v1/project/project-id: exited with 6
```

1. How to fix the dns issue?
2. Is it possible to add also-proxy config somehow via terminal. I want to provide a script for all developers, that they can develop a complex app locally like before with telepresence v1, without modifying local config files. I guess I could use a cli yaml parser to modify the kubeconfig but, well, this feels quite cumbersome.

**Comment by craftey** [comment]
> I do believe that you had to use `also-proxy` in v1 too, if you wanted access to non-cluster subnets. So perhaps there's more than one issue here?

In my initial post I showed the swap-deployment telepresence v1 cmd without also-proxy which works and I am trying to migrate to v2.

**Comment by thallgren** [comment]
The name `metadata.google.internal` is not handled by the Telepresence DNS out-of-the box, because `internal` is neither the name of a namespace, name of an intercepted service, or a top-level domain. Luckily, this can be fixed too, by using [includeSuffixes](https://www.getambassador.io/docs/telepresence/latest/reference/config/#dns). You'll need to add `.internal` or `.google.internal` to the list.

**Comment by DanielYWoo** [comment]
maybe we can have the traffic manager to detect a few env variables to detect gcp managed cluster and do that automatically
________________________________
发件人: Thomas Hallgren ***@***.***>
发送时间: Friday, October 7, 2022 7:04:18 PM
收件人: telepresenceio/telepresence ***@***.***>
抄送: Daniel Wu ***@***.***>; Mention ***@***.***>
主题: Re: [telepresenceio/telepresence] Migration from v1 to v2 not working, intercept not working (minimal example for GCP inside) (Issue #2798)


The name metadata.google.internal is not handled by the Telepresence DNS out-of-the box, because internal is neither the name of a namespace, name of an intercepted service, or a top-level domain. Luckily, this can be fixed too, by using includeSuffixes<https://www.getambassador.io/docs/telepresence/latest/reference/config/#dns>. You'll need to add .internal or .google.internal to the list.

―

**Comment by thallgren** [comment]
@DanielYWoo Indeed. That, and several other improvements will be included in the coming 2.8.0 release. There, `alsoProxy`, `neverProxy`, `includeSuffixes`, and `excludeSuffixes` can all be set as values in the Helm chart.

**Comment by craftey** [comment]
> The name `metadata.google.internal` is not handled by the Telepresence DNS out-of-the box, because `internal` is neither the name of a namespace, name of an intercepted service, or a top-level domain. Luckily, this can be fixed too, by using [includeSuffixes](https://www.getambassador.io/docs/telepresence/latest/reference/config/#dns). You'll need to add `.internal` or `.google.internal` to the list.

Thanks again @thallgren 

I added to the extension config in local kubeconfig:
```
        dns:
          include-suffixes:
          - metadata.google.internal
```
Now, DNS works. I will now try to my get my real app working with more external services than in this minimal example and report back.
(I understand that I could set the suffix more generic. I just tried this very narrow variant)

**Comment by craftey** [comment]
I got my real app working, by adding

```
    extensions:
    - extension:
        also-proxy:
        - 169.254.169.254/32 # metadata.google.internal
        - 10.xx.xx.0/26 # dev kafka
        - 10.xx.xx.xx/32 # vault.dev.myproject.com
        dns:
          include-suffixes:
          - metadata.google.internal
      name: telepresence.io
```
to the cluster entry in my `~/.kube/config`

I learned that since v1.8.x this can be fully configured via helm values for traffic-manager: 
```
telepresence helm install --values values.yaml
```
That is better then communicating this `~/.kube/config`-patch to all developers in my team. 

But still in telepresence-legacy this was not needed – no also-proxy, no include-suffixes – it just worked by swap-deployment.
@thallgren @cindymullins-dw
Can u comment on that last sentence please, will it ever be possible in v2 to have such a convenient config like with v1 swap-deployment? Maybe #2851 also goes in that direction?

Anyway, we can close this issue, because my technical questions are all solved.

Thanks a lot for your support!

**Comment by thallgren** [comment]
I know that also-proxy existed in v1 for the same purpose, but I don't know why you didn't need to use it. My guess is that with the v1 swap deployment, the local process never needed to communicate with the cluster resources from your workstation. Some DNS resolution and routing must take place for that to be possible, hence the also-proxy and include-suffixes.

**Comment by craftey** [comment]
@thallgren Yes it seems that the "proxy pod" that was swapped in place with v1 did all the magic for us. So it may have been communication between the local app and that proxy pod when talking to other resources not only in the cluster but in the cloud project, just everything the original pod had access to. It felt like we were running the local app directly in the remote cluster.

**Comment by thallgren** [comment]
My guess then, is that you used `--method=inject-tcp` for the intercept. It injects a shared-library into the sub-process that runs the interceptor. It's somewhat limited in that it doesn't work with docker containers, statically linked executables, suid executables, or if the process calls other processes, and IIRC it isn't available on Windows.

**Comment by craftey** [comment]
@thallgren no, I do not use `--method=inject-tcp`. The legacy command I use is shown in my initial post.

**Comment by thallgren** [comment]
That's odd, because it means that the default method `vpn-tcp` is used. The [docs](https://archive.getambassador.io/docs/telepresence/1.0/reference/methods/) say this about that method:
```
Cloud resources like AWS RDS will not be routed automatically via cluster. You'll need to specify the hosts manually using --also-proxy,
e.g. --also-proxy mydatabase.somewhere.vpc.aws.amazon.com to route traffic to that host via the Kubernetes cluster. Specify multiple
hosts to --also-proxy like --also-proxy example1.com --also-proxy example2.com --also-proxy example3.com.
```
which is very similar to what v2 does, albeit using include-suffixes and CIDRs instead of hostnames.

**Comment by craftey** [comment]
Thanks for the docs link. This made it more clear to me.

Initially I wrote, that with telepresence-legacy I found, that

```
telepresence --swap-deployment hello-server --docker-run --rm curlimages/curl:7.85.0 ...
```
works fine for a cloud resource outside the cluster like 169.254.169.254 (metadata.google.internal)
while
```
telepresence --run curl ...
#or
telepresence --swap-deployment hello-server --run curl ...
```
just timed out.

With your [methods docu link](https://archive.getambassador.io/docs/telepresence/1.0/reference/methods/) it is now clear why.
The two latter commands use method `vpn-tcp` while the docker-run cmd uses method `container`.
I could fix the two latter commands by adding parameter`--also-proxy 169.254.169.254`.

And following a link in the methods docu to the [docker tutorial in section #How it works](https://archive.getambassador.io/docs/telepresence/1.0/tutorials/docker/#how-it-works) explains why docker-run with telepresence-legacy does not need also-proxy configs:

> **How it works**
Telepresence will start a new proxy container and then call docker run with whatever arguments you pass to --docker-run to start a container that will have its networking proxied. **All networking is proxied**:
> - Outgoing to Kubernetes.
> - **Outgoing to cloud resources outside the cluster**
> - Incoming connections from the cluster to ports specified with --expose.
Volumes and environment variables from the remote Deployment are also available in the container

So, that second one in the list is the nice feature of telepresency-legacy I would like to see in v2 as well :)

However, I am happy to use version 2.9.x where I can use traffic-manager helm values to add my `alsoProxySubnets` and `includeSuffixes` and a little bug #2870 is fixed.

Thank you very much again for helping and discussing. Keep up the good work!
Kindly Craftey

**Comment by thallgren** [comment]
Thanks. I'll keep this open, because that second feature might be something we'll want to consider for v2.

**Comment by mleclercq** [comment]
I got the same issue and added the `also-proxy: 169.254.169.254/32` but I'm still getting a connection refused.

Note I'm running with `--docker`. Not sure if that changes something...

```
curl -v http://metadata.google.internal/computeMetadata/v1/project/project-id -H 'Metadata-Flavor: Google'
*   Trying 169.254.169.254:80...
* connect to 169.254.169.254 port 80 failed: Connection refused
* Failed to connect to metadata.google.internal port 80: Connection refused
* Closing connection 0
curl: (7) Failed to connect to metadata.google.internal port 80: Connection refused
```

So the DNS is not a problem. `metadata.google.internal` is properly resolved to 169.254.169.254, but the TCP connection doesn't seem to go through.

Here are the logs from `connector.log` when I run the curl:

```
2023-05-08 21:31:14.9068 debug   docker/playground/dns/Server : ServeDNS 27963 AAAA   metadata.google.internal.
2023-05-08 21:31:14.9078 debug   docker/playground/dns/Server : Lookup AAAA "metadata.google.internal.dyson-matthieu.svc.cluster.local."
2023-05-08 21:31:14.9070 debug   docker/playground/dns/Server : ServeDNS 58676 A      metadata.google.internal.
2023-05-08 21:31:14.9095 debug   docker/playground/dns/Server : Lookup A "metadata.google.internal.dyson-matthieu.svc.cluster.local."
2023-05-08 21:31:14.9354 debug   docker/playground/dns/Server : Lookup AAAA "metadata.google.internal."
2023-05-08 21:31:14.9354 debug   docker/playground/dns/Server : Lookup A "metadata.google.internal."
2023-05-08 21:31:14.9655 debug   docker/playground/dns/Server : 27963 AAAA   metadata.google.internal. -> NOERROR EMPTY
2023-05-08 21:31:14.9691 debug   docker/playground/dns/Server : 58676 A      metadata.google.internal. -> NOERROR [169.254.169.254]
```

When I curl a kubernetes service, it works as expected with the following logs:

```
2023-05-08 21:31:38.4045 debug   docker/playground/dns/Server : ServeDNS 35922 AAAA   couchdb.
2023-05-08 21:31:38.4054 debug   docker/playground/dns/Server : Lookup AAAA "couchdb.dyson-matthieu.svc.cluster.local."
2023-05-08 21:31:38.4045 debug   docker/playground/dns/Server : ServeDNS 49748 A      couchdb.
2023-05-08 21:31:38.4071 debug   docker/playground/dns/Server : Lookup A "couchdb.dyson-matthieu.svc.cluster.local."
2023-05-08 21:31:38.4336 debug   docker/playground/dns/Server : 49748 A      couchdb. -> NOERROR [172.25.14.132]
2023-05-08 21:31:38.4336 debug   docker/playground/dns/Server : Lookup AAAA "couchdb."
2023-05-08 21:31:38.4552 debug   docker/playground/dns/Server : 35922 AAAA   couchdb. -> NOERROR EMPTY
2023-05-08 21:31:38.4599 debug   docker/playground/network : Opening tunnel for id tcp 172.25.0.0:53190 -> 172.25.14.132:5984
```

Here we have an additional `Opening tunnel for id tcp ...` log.

Also in the logs, when the connector starts, I see:

```
2023-05-08 20:31:38.1558 info    docker/playground/network : also-proxy subnets [169.254.169.254/32]
...
2023-05-08 20:31:38.1788 info    docker/playground/network : started command ["ip" "a" "add" "169.254.169.254/32" "dev" "tel0"] : dexec.pid="39"
2023-05-08 20:31:38.1796 info    docker/playground/network :  : dexec.pid="39" dexec.stream="stdin" dexec.err="EOF"
2023-05-08 20:31:38.1801 info    docker/playground/network : finished successfully: exit status 0 : dexec.pid="39"
```

So looks like the `also-proxy: 169.254.169.254/32` is taken into account.

When I "docker exec" in the daemon container, I get the following ip config:

```
/ # ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
2: tunl0@NONE: <NOARP> mtu 1480 qdisc noop state DOWN qlen 1000
    link/ipip 0.0.0.0 brd 0.0.0.0
3: ip6tnl0@NONE: <NOARP> mtu 1452 qdisc noop state DOWN qlen 1000
    link/tunnel6 00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00 brd 00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00
4: tel0: <POINTOPOINT,MULTICAST,NOARP,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UNKNOWN qlen 500
    link/[65534] 
    inet 172.25.0.0/20 scope global tel0
       valid_lft forever preferred_lft forever
    inet 10.105.0.0/23 scope global tel0
       valid_lft forever preferred_lft forever
    inet 169.254.169.254/32 scope global tel0
       valid_lft forever preferred_lft forever
77: eth0@if78: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1500 qdisc noqueue state UP 
    link/ether 02:42:ac:12:00:02 brd ff:ff:ff:ff:ff:ff
    inet 172.18.0.2/16 brd 172.18.255.255 scope global eth0
       valid_lft forever preferred_lft forever
/ # 
/ # 
/ # ip r
default via 172.18.0.1 dev eth0 
10.105.0.0/23 dev tel0 scope link  src 10.105.0.0 
172.18.0.0/16 dev eth0 scope link  src 172.18.0.2 
172.25.0.0/20 dev tel0 scope link  src 172.25.0.0 
```

So the `169.254.169.254/32` IP is there on the `tel0` interface but there is no route. Not sure if one is needed...

**Comment by thallgren** [comment]
Using a submask that is larger than 30 bits doesn't seem to work for the TUN-device when used in docker. We'll need to investigate why that is.
